src/backend/distributed/utils/shardinterval_utils.c

```diff
@@ -0,0 +1,293 @@
+/*-------------------------------------------------------------------------
+ *
+ * shardinterval_utils.c
+ *
+ * This file contains functions to perform useful operations on shard intervals.
+ *
+ * Copyright (c) 2014-2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include "access/nbtree.h"
+#include "catalog/pg_am.h"
+#include "catalog/pg_collation.h"
+#include "catalog/pg_type.h"
+#include "distributed/shardinterval_utils.h"
+#include "distributed/pg_dist_partition.h"
+#include "distributed/worker_protocol.h"
+#include "utils/catcache.h"
+#include "utils/memutils.h"
+
+
+/*
+ * ShardIntervalCompareFunctionCache is used for caching shard interval
+ * compare functions.
+ */
+static List *ShardIntervalCompareFunctionCache = NIL;
+
+
+static FmgrInfo * ShardIntervalCompareFunction(Var *partitionColumn,
+											   char partitionMethod);
+static ShardInterval * SearchCachedShardInterval(Datum partitionColumnValue,
+												 ShardInterval **shardIntervalCache,
+												 int shardCount,
+												 FmgrInfo *compareFunction);
+
+
+/*
+ * CompareShardIntervals acts as a helper function to compare two shard intervals
+ * by their minimum values, using the value's type comparison function.
+ *
+ * If a shard interval does not have min/max value, it's treated as being greater
+ * than the other.
+ */
+int
+CompareShardIntervals(const void *leftElement, const void *rightElement,
+					  FmgrInfo *typeCompareFunction)
+{
+	ShardInterval **leftShardInterval = (ShardInterval **) leftElement;
+	ShardInterval **rightShardInterval = (ShardInterval **) rightElement;
```

Comment:
Given that you're only dealing with dereferenced ShardInterval **'s in the rest of the function, maybe dereference them here? I think that'd make the code easier to read.